state/protocol/badger/mutator.go

```diff
@@ -322,6 +322,12 @@ func (m *MutableState) guaranteeExtend(ctx context.Context, candidate *flow.Bloc
 			return state.NewInvalidExtensionErrorf("payload includes expired guarantee (height: %d, limit: %d)",
 				ref.Height, limit)
 		}
+
+		// check the guarantors are correct
+		_, err = protocol.FindGuarantors(m, guarantee)
+		if err != nil {
```

Comment:
## ⚠️ 

It would be very important to me to consequently avoid this pattern of "seeing an error and attributing it without further consideration to one particular root cause". 

Reasoning: 
* Consider the case of a corrupted data base, returning some error on read. Or consider a bug in some of the code, where `FindGuarantors` returns an unexpected error. Your code would interpret any error as a symptom of the block being invalid. 
    * At some point in the not-too-distant future, this will result in a slashing challenge against the block proposer. If this accusation is false, the node raising an invalid challenge will be slashed (causing hundreds of thousands of dollars in damage)! 
    * Lets anticipate the point in time when we want to start slashing challenges. Do we want to go over every line of code and look for cases like this? I think that would be prohibitively expensive. 
* In addition, I think this type of undifferentiated error attribution also makes our code very brittle. Maybe at the moment, any error might be a symptom of a broken input. But some engineer might extend the logic of `FindGuarantors` in the future (maybe adding an additional sanity check). 
   * I think it would not be a reasonable ask that an engineer has to go through all usages of the function they extended with additional sanity checks and see whether they broke some implicit assumptions that go beyond the documented API. 
   * The API of `FindGuarantors` is: https://github.com/onflow/flow-go/blob/8c5e6ab0d29f46b4ea954fb2d7e2a40d74947998/state/protocol/util.go#L75-L79 This means that the sentinel errors (and only these!) indicate potentially byzantine inputs  `ErrIncompatibleBitVectorLength`, `ErrIllegallyPaddedBitVector`, `ErrInvalidChecksum`. Other errors are _not_ to be interpreted as symptom of a byzantine input. 


The unfortunate reality is that the protocol state does not properly differentiate between byzantine inputs and unexpected errors. **My ask to you**:
* For the code you add (or the parts of the code you touch), please add byzantine-compatible error handling. Please don't exacerbate the problem by taking the absence of byzantine-compatible error handling in existing code as a reason to ignore it in new code. 

### Problem:
There is a circular assumption in the code regarding detection of byzantine cases, which is not satisfied. Assume a byzantine input, where the `guarantee.ChainID` has some garbage value. 
* `FindGuarantors` delegates the responsibility of detecting this case to the protocol state: https://github.com/onflow/flow-go/blob/8c5e6ab0d29f46b4ea954fb2d7e2a40d74947998/state/protocol/util.go#L86-L87 
* And the protocol state assumes that `FindGuarantors` will spot this problem. 

I am hoping that by advocating for detailed considerations of byzantine inputs, we would increase our chances of finding such problems, se we create less technical debt for later. 

_Suggestion_
1. review the implementations of `Epoch`'s `ClusterByChainID` method and figure out what sentinel errors are expected during normal operations 
2. Document the expected sentinel errors in the [`Epoch` interface](https://github.com/onflow/flow-go/blob/8c5e6ab0d29f46b4ea954fb2d7e2a40d74947998/state/protocol/epoch.go#L80-L82)
3. `FindGuarantors` should handle the expected errors [here](https://github.com/onflow/flow-go/blob/8c5e6ab0d29f46b4ea954fb2d7e2a40d74947998/state/protocol/util.go#L85) accordingly. The goDoc should clearly specify what sentinel is returned in case the cluster is unknown.
4. Differentiate between expected errors and unexpected exceptions in here in line 328 of the `MutableState`.  Something like the following would be super great:
  ```golang
  if err != nil {
     if errors.Is(err, signature.ErrIllegallyPaddedBitVector) || errors.Is(err, signature.ErrIncompatibleBitVectorLength) || errors.Is(err, signature.ErrInvalidChecksum) ||  errors.Is(err, protocol.ErrClusterNotFound) {
         return state.NewInvalidExtensionErrorf("invalid collection guarantee %v: %w", guarantee.ID(), err)
     }
     return fmt.Errorf("unexpected internal error validating guarantors of collection %v: %w", guarantee.ID(), err)
  }
  ```

As part of your PR I would suggest to not touch anything at a higher level, i.e. do not touch the error handling in `MutableState` beyond extending line 328. But at least with this bottom-up approach, we are incrementally adding proper error piece by piece.

Reply:
Good point.

It is more and more common that we would like to distinguish expected error and unexpected exceptions. What you suggested works, but with so many checks in the if statement: `if errors.Is(er, sentinelError1) || errors.Is(err, sentinelError2) || ...)`, it's not easy to maintain. Adding one more sentinel error would have to update all places.

I have two options:
1. Add a `IsUnexpectedFindGuarantorsError(err error) bool` function 
2. Let FindGuarantors to return an extra bool to indicate whether the error is an exception. This could remind the caller to distinguish exception.

1 is simpler, but 2 is more explicit.

Reply:
In my opinion, find `FindGuarantors` cannot decide in all cases, whether an error is expected or unexpected. A good example is the problem at hand:
* When extending the protocol state with a new block, we validate the collection guarantees. A collection guarantee including an unknown value for `guarantee.ChainID` is expected because the block could be proposed by a byzantine consensus node. 
* In contrast, when the execution node decodes the guarantors, the block was already approved by the consensus nodes. Finding an unknown value for `guarantee.ChainID` is fatal in this case.

Reply:
Updated the error handling to handle each sentinel errors separately.